File: android/core/hangul/src/main/kotlin/app/piyokey/core/hangul/OSIMETextJudge.kt
package app.piyokey.core.hangul

sealed interface OSIMETextJudgeStatus {
  /** The full visible text is a valid target prefix. Marked text may still change. */
  data class Matching(
    val completed: Boolean,
    val isComposing: Boolean,
  ) : OSIMETextJudgeStatus

  /** Only the marked (unconfirmed) portion differs, so no mistake is recorded. */
  data object ComposingMismatch : OSIMETextJudgeStatus

  /** Confirmed text differs from the target and counts as one mistake. */
  data class ConfirmedMismatch(val expectedIndex: Int) : OSIMETextJudgeStatus
}

data class OSIMETextEvaluation(
  val status: OSIMETextJudgeStatus,
  /** The longest safe target prefix that the session may reflect immediately. */
  val acceptedSequence: List<Char>,
)

object OSIMETextJudge {
  /**
   * Judges text received from an OS IME against a target's two-beolsik jamo sequence.
   *
   * [committedText] excludes the active marked range. [markedText] is the active composition at
   * the end of that text, if any. A mismatching marked range is ignored until the IME confirms it.
   */
  fun evaluate(
    target: String,
    committedText: String,
    markedText: String? = null,
  ): OSIMETextEvaluation {
    val expected = JamoDecomposer.keySequenceFor(target)
    val committed = keySequenceAllowingEmptyOrNull(committedText)

    if (committed == null) {
      val validPrefix = longestDecomposablePrefix(committedText)
      if (validPrefix.size >= expected.size && validPrefix.take(expected.size) == expected) {
        return OSIMETextEvaluation(
          status = OSIMETextJudgeStatus.Matching(completed = true, isComposing = false),
          acceptedSequence = expected,
        )
      }
      val mismatch = mismatchIndex(validPrefix, expected) ?: validPrefix.size
      return OSIMETextEvaluation(
        status = OSIMETextJudgeStatus.ConfirmedMismatch(expectedIndex = mismatch),
        acceptedSequence = expected.take(mismatch),
      )
    }

    if (committed.size >= expected.size && committed.take(expected.size) == expected) {
      return OSIMETextEvaluation(
        status = OSIMETextJudgeStatus.Matching(completed = true, isComposing = false),
        acceptedSequence = expected,
      )
    }

    val committedMismatch = mismatchIndex(committed, expected)
    if (committedMismatch == null) {
      if (!markedText.isNullOrEmpty()) {
        val marked = keySequenceAllowingEmptyOrNull(markedText)
          ?: return OSIMETextEvaluation(
            status = OSIMETextJudgeStatus.ComposingMismatch,
            acceptedSequence = committed,
          )
        val visible = committed + marked
        if (mismatchIndex(visible, expected) != null) {
          return OSIMETextEvaluation(
            status = OSIMETextJudgeStatus.ComposingMismatch,
            acceptedSequence = committed,
          )
        }
        return OSIMETextEvaluation(
          status = OSIMETextJudgeStatus.Matching(
            completed = visible.size == expected.size,
            isComposing = true,
          ),
          acceptedSequence = visible,
        )
      }

      return OSIMETextEvaluation(
        status = OSIMETextJudgeStatus.Matching(
          // Exact completion returned above, so a matching committed prefix is incomplete here.
          completed = false,
          isComposing = false,
        ),
        acceptedSequence = committed,
      )
    }

    return OSIMETextEvaluation(
      status = OSIMETextJudgeStatus.ConfirmedMismatch(expectedIndex = committedMismatch),
      acceptedSequence = expected.take(committedMismatch),
    )
  }

  private fun keySequenceAllowingEmptyOrNull(text: String): List<Char>? = try {
    if (text.isEmpty()) emptyList() else JamoDecomposer.keySequenceFor(text)
  } catch (_: JamoDecompositionException) {
    null
  }

  private fun mismatchIndex(candidate: List<Char>, expected: List<Char>): Int? {
    candidate.forEachIndexed { index, input ->
      if (index >= expected.size || input != expected[index]) return index
    }
    return null
  }

  private fun longestDecomposablePrefix(text: String): List<Char> = buildList {
    for (character in text) {
      val sequence = try {
        JamoDecomposer.keySequenceFor(character.toString())
      } catch (_: JamoDecompositionException) {
        break
      }
      addAll(sequence)
    }
  }
}

File: ios/HangulEngine/Sources/HangulEngine/OSIMETextJudge.swift
public enum OSIMETextJudgeStatus: Equatable, Sendable {
  /// The full visible text is a valid target prefix. Marked text may still change.
  case matching(completed: Bool, isComposing: Bool)
  /// Only the marked (unconfirmed) portion differs, so no mistake is recorded.
  case composingMismatch
  /// An ASCII key arrived from a non-Korean input source and must not affect the session.
  case unsupportedASCIIInput
  /// Confirmed text differs from the target and counts as one mistake.
  case confirmedMismatch(expectedIndex: Int)
}

public struct OSIMETextEvaluation: Equatable, Sendable {
  public let status: OSIMETextJudgeStatus
  /// The longest safe target prefix that the session may reflect immediately.
  public let acceptedSequence: [Character]

  public init(status: OSIMETextJudgeStatus, acceptedSequence: [Character]) {
    self.status = status
    self.acceptedSequence = acceptedSequence
  }
}

public enum OSIMETextJudge {
  /// Judges text received from an OS IME against a target's two-beolsik jamo sequence.
  ///
  /// `committedText` excludes the active marked range. `markedText` is the active
  /// composition at the end of that text, if any. A mismatching marked range is
  /// intentionally ignored until the IME confirms it.
  public static func evaluate(
    target: String,
    committedText: String,
    markedText: String? = nil
  ) throws -> OSIMETextEvaluation {
    let expected = try JamoDecomposer.keySequence(for: target)
    if containsUnsupportedASCII(in: committedText) || containsUnsupportedASCII(in: markedText ?? "") {
      let validPrefix = longestDecomposablePrefix(of: committedText)
      if validPrefix.count >= expected.count,
        Array(validPrefix.prefix(expected.count)) == expected
      {
        return OSIMETextEvaluation(
          status: .matching(completed: true, isComposing: false),
          acceptedSequence: expected
        )
      }
      let acceptedCount = mismatchIndex(candidate: validPrefix, expected: expected)
        ?? min(validPrefix.count, expected.count)
      return OSIMETextEvaluation(
        status: .unsupportedASCIIInput,
        acceptedSequence: Array(expected.prefix(acceptedCount))
      )
    }

    guard let committed = try? keySequenceAllowingEmpty(for: committedText) else {
      let validPrefix = longestDecomposablePrefix(of: committedText)
      if validPrefix.count >= expected.count,
        Array(validPrefix.prefix(expected.count)) == expected
      {
        return OSIMETextEvaluation(
          status: .matching(completed: true, isComposing: false),
          acceptedSequence: expected
        )
      }
      let mismatch = mismatchIndex(candidate: validPrefix, expected: expected) ?? validPrefix.count
      return OSIMETextEvaluation(
        status: .confirmedMismatch(expectedIndex: mismatch),
        acceptedSequence: Array(expected.prefix(mismatch))
      )
    }

    if committed.count >= expected.count,
      Array(committed.prefix(expected.count)) == expected
    {
      return OSIMETextEvaluation(
        status: .matching(completed: true, isComposing: false),
        acceptedSequence: expected
      )
    }

    guard let committedMismatch = mismatchIndex(candidate: committed, expected: expected) else {
      if let markedText, !markedText.isEmpty {
        guard let marked = try? keySequenceAllowingEmpty(for: markedText) else {
          return OSIMETextEvaluation(
            status: .composingMismatch,
            acceptedSequence: committed
          )
        }
        let visible = committed + marked
        guard mismatchIndex(candidate: visible, expected: expected) == nil else {
          return OSIMETextEvaluation(
            status: .composingMismatch,
            acceptedSequence: committed
          )
        }
        return OSIMETextEvaluation(
          status: .matching(completed: visible.count == expected.count, isComposing: true),
          acceptedSequence: visible
        )
      }

      return OSIMETextEvaluation(
        status: .matching(completed: committed.count == expected.count, isComposing: false),
        acceptedSequence: committed
      )
    }

    return OSIMETextEvaluation(
      status: .confirmedMismatch(expectedIndex: committedMismatch),
      acceptedSequence: Array(expected.prefix(committedMismatch))
    )
  }

  private static func keySequenceAllowingEmpty(for text: String) throws -> [Character] {
    text.isEmpty ? [] : try JamoDecomposer.keySequence(for: text)
  }

  private static func containsUnsupportedASCII(in text: String) -> Bool {
    text.unicodeScalars.contains { scalar in
      scalar.isASCII && scalar.value != 0x20
    }
  }

  private static func mismatchIndex(
    candidate: [Character],
    expected: [Character]
  ) -> Int? {
    for (index, input) in candidate.enumerated() {
      guard index < expected.count, input == expected[index] else { return index }
    }
    return nil
  }

  private static func longestDecomposablePrefix(of text: String) -> [Character] {
    var result: [Character] = []
    for character in text {
      guard let sequence = try? JamoDecomposer.keySequence(for: String(character)) else {
        break
      }
      result.append(contentsOf: sequence)
    }
    return result
  }
}
